src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -1712,23 +1720,24 @@ bool StorageReplicatedMergeTree::tryExecuteMerge(const LogEntry & entry)
     future_merged_part.updatePath(*this, reserved_space);
     future_merged_part.merge_type = entry.merge_type;
 
-    if (storage_settings_ptr->allow_s3_zero_copy_replication)
+    auto disk_type = reserved_space->getDisk()->getType();
+    if ((disk_type == DiskType::Type::S3
+        && storage_settings_ptr->allow_s3_zero_copy_replication
+        && merge_strategy_picker.shouldMergeOnSingleReplicaS3Shared(entry))
+        || (disk_type == DiskType::Type::HDFS
+        && storage_settings_ptr->allow_hdfs_zero_copy_replication
+        && merge_strategy_picker.shouldMergeOnSingleReplicaHdfsShared(entry)))
     {
-        if (auto disk = reserved_space->getDisk(); disk->getType() == DB::DiskType::Type::S3)
-        {
-            if (merge_strategy_picker.shouldMergeOnSingleReplicaS3Shared(entry))
-            {
-                if (!replica_to_execute_merge_picked)
-                    replica_to_execute_merge = merge_strategy_picker.pickReplicaToExecuteMerge(entry);
+        if (!replica_to_execute_merge_picked)
+            replica_to_execute_merge = merge_strategy_picker.pickReplicaToExecuteMerge(entry);
 
-                if (replica_to_execute_merge)
-                {
-                    LOG_DEBUG(log,
-                        "Prefer fetching part {} from replica {} due s3_execute_merges_on_single_replica_time_threshold",
-                        entry.new_part_name, replica_to_execute_merge.value());
-                    return false;
-                }
-            }
+        if (replica_to_execute_merge)
+        {
+            const auto * param = (disk_type == DiskType::Type::S3) ? "s3_execute_merges_on_single_replica_time_threshold" : "hdfs_execute_merges_on_single_replica_time_threshold";
```

Comment:
I'm agree, but how we should name it? `external_storage_execute_merges_on_single_replica`?

Reply:
Since now disks `s3` and `hdfs` are united under base class IDiskRemote, and it uses flags like `keep_in_remote_fs` (in `removeShared` btw), then may be `remote_fs_execute_merges_on_single_replica`?

Reply:
cc @ianton-ru, there are settings changes

Reply:
I vote for one setting name for all types of remote storages.